internal/pkg/log/log.go

```diff
@@ -2,105 +2,123 @@
 package log
 
 import (
-	"io"
-	"os"
+	"context"
 
-	"go.uber.org/zap"
-	"go.uber.org/zap/zapcore"
+	"github.com/envoyproxy/xds-relay/internal/pkg/log/zap"
+
+	z "go.uber.org/zap"
 )
 
-// Options contains all possible log settings.
-type Options struct {
-	// Level configures the log verbosity. Defaults to Debug.
-	Level *zap.AtomicLevel
-	// StacktraceLevel is the level which stacktraces will be emitted. Defaults
-	// to Warn.
-	StacktraceLevel *zap.AtomicLevel
-	// Encoder configures how Zap will encode the output. Defaults to JSON.
-	Encoder zapcore.Encoder
-	// OutputDest controls the destination of the log output. Defaults to
-	// os.Stderr.
-	OutputDest io.Writer
-	// ZapOptions allows passing additional optional zap.Options, ex: Sampling.
-	ZapOptions []zap.Option
+// Logger is the contract for xds-relay's logging implementation.
+//
+// A self-contained usage example looks as follows:
+//
+//    Log.Named("foo-component").With(
+//        "field1", "value1",
+//        "field2", "value2",
+//     ).Error("my error message")
+type Logger interface {
+	// Named adds a sub-scope to the logger.
+	Named(name string) Logger
+
+	// With adds a variadic number of fields to the logging context.
+	// When processing pairs, the first element of the pair is used as the
+	// field key and the second as the field value.
+	//
+	// For example,
+	//
+	//   Log.With(
+	//     "hello", "world",
+	//     "failure", errors.New("oh no"),
+	//     "count", 42,
+	//     "user", User{Name: "alice"},
+	//  ).Info("this is an error message")
+	With(args ...interface{}) Logger
+
+	// Log a message at level Debug, annotated with fields provided through With().
+	Debug(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Info, annotated with fields provided through With().
+	Info(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Warn, annotated with fields provided through With().
+	Warn(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Error, annotated with fields provided through With().
+	Error(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Panic, annotated with fields provided through With(), and immediately
+	// panic.
+	Panic(ctx context.Context, msg ...interface{})
+
+	// Log a message at level Fatal, annotated with fields provided through With(), and immediately
+	// call os.Exit.
+	Fatal(ctx context.Context, msg ...interface{})
+
+	// Sync flushes any buffered log entries.
+	Sync() error
```

Comment:
how is this going to be called?

Reply:
We'll be calling it in the shutdown handlers of the main gRPC server, https://godoc.org/go.uber.org/zap#hdr-Choosing_a_Logger